File: golang/valid_parentheses.go
func isValid(s string) bool {
	stack := []rune{}
	for _, char := range s {
		if len(stack) == 0 {
			stack = append(stack, char)
		} else {
			if (stack[len(stack)-1] == '(' && char == ')') || (stack[len(stack)-1] == '{' && char == '}') || (stack[len(stack)-1] == '[' && char == ']') {
				stack = stack[:len(stack)-1]
			} else {
				stack = append(stack, char)
			}
		}
	}
	return len(stack) == 0
}

File: typescript/valid_parentheses.ts
function isValid(s: string): boolean {
	let stack = []

	for (let char of s) {
		if (char == '(' || char == '{' || char == '[') {
			stack.push(char)
		} else if (char == ')' || char == '}' || char == ']') {
			if (stack.length == 0) return false
			let top = stack.pop()
			if (top != '(' && char == ')') return false
			if (top != '{' && char == '}') return false
			if (top != '[' && char == ']') return false
		}
	}

	return stack.length == 0
}
